source/extensions/filters/http/ext_proc/ext_proc.cc

```diff
@@ -397,10 +397,12 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
     // may request a watermark if the queue is higher than the buffer limit to prevent running
     // out of memory.
     // 2) As a result, filters farther down the chain see empty buffers in some data callbacks.
-    // 3) When a response comes back from the external processor, it injects the processor's result
-    // into the filter chain using "inject**codedData". (The processor may respond indicating that
-    // there is no change, which means that the original buffer stored in the queue is what gets
-    // injected.)
+    // 3) When a response comes back from the external processor, it is buffered in HCM's buffered
+    // data and will be propagated to further filters in the chain altogether later (The processor
+    // may respond indicating that there is no change, which means that the original buffer stored
+    // in the queue is what gets propagated.)
+    // 4) If the response from the external processor is too large, watermark and then local reply
```

Comment:
How would someone write an ext_proc service that decompresses the message and needs to be able to send the data mutation larger than the original message?

Reply:
They still can send the data mutation larger than the original message, but just not unlimited amplification now. 
The response data that is above watermark buffer limit will be rejected.

Reply:
In short, there is a trade-off between OOM risk and flexibility/customer experience.  Customer still can decompress/amplify the message but now there is limitation (which is configurable via `buffer_limit_`).

Most ideal solution is applying back pressure on sidestream. It has been explored and it requires significant effort. Meanwhile, we want to resolve OOM risk first.  
cc @htuch , we have briefly discussed this last week.

Reply:
This design prevents even smallest - 1 byte - amplification. If Envoy received large data - the entire buffer size - and ext_proc sends it to the server, the ext_proc server can not add even 1 byte without causing the request to fail.

Reply:
Yes, when original data is already at buffer limit, amplification will not be allowed here.  I think from OOM perspective, such small amplification is same as huge amplification to large response. The end result is huge data injected into Envoy, which leads to OOM risk.

Reply:
Why is small amplification the same as huge amplification? There are a lot of use cases where increase in body size is expected. There are existing use case where ext_proc examines and modifies proto in the gRPC request. This can easily lead to making data slightly larger. We should allow increases to body size, otherwise it limits the usefulness of ext_proc. This change prevents any possibility of increasing body size.

Note that the core of the problem is not the increase in the body size, but that the larger body is sent in **one** response from ext_proc. To avoid this, we need to allow ext_proc to send larger body in multiple responses.

Reply:
i think `allow ext_proc to send larger body in multiple responses` will not resolve the issue. OOM can be caused by either(1) single large response or (2)multiple smaller responses in short period that mainstream can not process.  The end result is that data is back up in mainstream.  

We can implement multiple response. But same as max_gRPC_length limitation it is not enough to prevent OOM. I agree that this is short-term solution for OOM while long-term solution that sidestream flow control is being developed

Reply:
Yes, flow control on sidestream is long-term solution (as described in PR description). 
It is being actively explored/developed but it requires careful design and significant effort that can not be available shortly.